### PR TITLE
Add caching for Pen/SoldBrush

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Adds base classes and caches for System.Drawing pens and brushes. Signficantly reduces allocations for Graphics drawing.

- Remove other instances of Brush/Pen caching
- Write helpers to retain the orginal Color when finding nearest color (allowing existing caches to be used)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3644)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/SinglyLinkedList.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                    result = false;
+                }
+                else if (_current is null)
+                {
+                    // At the start
+                    _current = _list.First;
+                }
+                else if (_current.Next != null)
+                {
+                    // Not to the end yet
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/SinglyLinkedList.cs:124]:
```suggestion
                    // Not at the end yet
```

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/SystemDrawingExtensions.cs [hunk 2]
```diff
@@ -15,8 +15,36 @@ internal static class SystemDrawingExtensions
     {
         internal static Gdi32.HBITMAP GetHBITMAP(this Bitmap bitmap) => (Gdi32.HBITMAP)bitmap.GetHbitmap();
         internal static Gdi32.HFONT ToHFONT(this Font font) => (Gdi32.HFONT)font.ToHfont();
+
+        /// <summary>
+        ///  Similar to <see cref="Graphics.GetNearestColor(Color)"/>, but this retains the original color if the color
+        ///  didn't actually change. This retains the state of the color.
+        /// </summary>
+        /// <remarks>
+        ///  This is important as the color only changes if <paramref name="graphics"/> is a very low color depth. This
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/SystemDrawingExtensions.cs:24]:
```suggestion
        ///  This is important as the color only changes if <paramref name="graphics"/> has a very low color depth. This
```

File: src/System.Windows.Forms.Primitives/tests/UnitTests/System/Windows/Forms/RefCacheTests.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            using var seventhScope = cache.GetEntry(7);
+
+            // Now that the other scopes are closed, the earliest entries should have been cleaned
+            Assert.Equal(1, firstScope.Object.DisposeCount);
+            Assert.Equal(1, secondScope.Object.DisposeCount);
+            Assert.Equal(0, thirdScope.Object.DisposeCount);
+            Assert.Equal(0, fourthScope.Object.DisposeCount);
+        }
+
+        internal sealed class IntColorCache : RefCountedCache<Color, Color, int>
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/tests/UnitTests/System/Windows/Forms/RefCacheTests.cs:80]: Is this used?

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs [hunk 4]
```diff
@@ -16668,12 +16601,18 @@ protected override void OnPaint(PaintEventArgs e)
 
                 if (clipRect.IntersectsWith(gridRect))
                 {
-                    using (Region clipRegion = g.Clip)
+                    using (Region originalClip = g.Clip)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:16604]: Why not use `GraphicsClipScope`?

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/RefCountedCache.Scope.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                Entry?.RemoveRef();
+                SuppressFinalize();
+            }
+
+            [Conditional("DEBUG")]
+            public void SuppressFinalize()
+            {
+                GC.SuppressFinalize(this!);
+            }
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/RefCountedCache.Scope.cs:84]: Why not move this into `DisposalTracking` class?

File: src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.Polygon.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
[the remark this answers is not in the document]
Comment (weltkante) [reply src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.Polygon.cs:1]: That's a weird spelling? `polygon` is just one word, so camel case makes not much sense?

[Docs](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-polygon) also spell the function `Polygon` so maybe this is just a typo?

[the remark this answers is not in the document]
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.Polygon.cs:1]: Most probably, I didn't check the docs.

File: src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetWindowRect.cs [hunk 7]
```diff
@@ -11,7 +11,7 @@ internal static partial class Interop
     internal static partial class User32
     {
         [DllImport(Libraries.User32, ExactSpelling = true)]
-        public static extern BOOL GetWindowRect(IntPtr hWnd, ref RECT rect);
+        public unsafe static extern BOOL GetWindowRect(IntPtr hWnd, ref RECT rect);
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.GetWindowRect.cs:14]: It isn't, I backpedaled here but not completely. Removing.

File: src/System.Windows.Forms.Primitives/tests/UnitTests/System/Windows/Forms/RefCacheTests.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/tests/UnitTests/System/Windows/Forms/RefCacheTests.cs:80]: Not currently, it is there as an example of an alternative way to configure the cache for now. Added comments explaining.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:16604]: Overlooked it. Fixed.

File: src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.Polygon.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/Interop/Gdi32/Interop.Polygon.cs:1]: Polygon is correct in this case

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/RefCountedCache.Scope.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/RefCountedCache.Scope.cs:84]: I've moved it up as a static and call it like this:

``` C#
DisposalTracking.SuppressFinalize(this!);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Internal/Gdi/FontCache.Data.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                    lfOutPrecision = Gdi32.OUT_PRECIS.TT,
+                    lfQuality = quality,
+                    lfWeight = (font.Style & FontStyle.Bold) == FontStyle.Bold ? Gdi32.FW.BOLD : Gdi32.FW.NORMAL,
+                    lfItalic = (font.Style & FontStyle.Italic) == FontStyle.Italic ? True : False,
+                    lfUnderline = (font.Style & FontStyle.Underline) == FontStyle.Underline ? True : False,
+                    lfStrikeOut = (font.Style & FontStyle.Strikeout) == FontStyle.Strikeout ? True : False,
+                    FaceName = familyName
+                };
+
+                if (logFont.FaceName.IsEmpty)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Internal/Gdi/FontCache.Data.cs:105]: Since `LOGFONTW.FaceName` is a span, should we check for and strip vertical-font mark from the name if needed on the span instead re-allocating string on line:74?

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        [ThreadStatic]
+        private static SolidBrushCache? s_brushCache;
+
+        // Picking soft and hard limits expecting that 30 of each is a reasonable upper limit of permanently reserved
+        // space. These items are roughly .5KB each, with _most_ of the memory being native (in the GDI+ object).
+
+        private static PenCache PenCache => s_penCache ??= new PenCache(softLimit: 30, hardLimit: 40);
+        private static SolidBrushCache BrushCache => s_brushCache ??= new SolidBrushCache(softLimit: 30, hardLimit: 40);
+
+        private static PenCache.Scope GetPen(Color color)
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs:28]: This bothers me, for API readability and clarity of the intent should we add `Scope` suffix?
```suggestion
        private static PenCache.Scope GetPenScope(Color color)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 10]
```diff
@@ -8434,139 +8454,127 @@ private static void PaintBackColor(PaintEventArgs e, Rectangle rectangle, Color
 
             // Note: PaintEvent.HDC == 0 if GDI+ has used the HDC -- it wouldn't be safe for us
             // to use it without enough bookkeeping to negate any performance gain of using GDI.
-            if (color.A == 255)
+            if (color.HasTransparency())
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:8457]: I don't think it is equivalent, `A==255` is opaque
```suggestion
            if (!color.HasTransparency())
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 11]
```diff
@@ -8434,139 +8454,127 @@ private static void PaintBackColor(PaintEventArgs e, Rectangle rectangle, Color
...
+        /// <param name="rectangle">The area to redraw.</param>
+        /// <param name="transparentRegion">
+        ///  Region of the rectangle to be transparent, or null for the entire control.
+        /// </param>
         internal unsafe void PaintTransparentBackground(PaintEventArgs e, Rectangle rectangle, Region transparentRegion)
         {
             Control parent = ParentInternal;
 
-            if (parent != null)
+            if (parent == null)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:8512]:
```suggestion
            if (parent is null)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewComboBoxCell.cs [hunk 12]
```diff
@@ -2165,65 +2167,20 @@ private bool OwnsEditingComboBox(int rowIndex)
...
-                                if (highContrast)
-                                {
-                                    buttonShadowDark = ControlPaint.LightLight(color);
-                                }
-                                else
-                                {
-                                    buttonShadowDark = ControlPaint.DarkDark(color);
-                                }
-                            }
+                            Pen pen = SystemInformation.HighContrast ? SystemPens.ControlLight : SystemPens.Control;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewComboBoxCell.cs:2173]: It is really hard to parse what's changed here, but it feels some code paths are lost.
The original code had a tri-state for getting pens:
```cs
                            bool stockColor = color.ToKnownColor() == SystemColors.Control.ToKnownColor();
                            if (stockColor)	
                            {	
                                if (SystemInformation.HighContrast)	
                                {	
                                    pen = SystemPens.ControlLight;	
                                }	
                                else	
                                {	
                                    pen = SystemPens.Control;	
                                }	
                            }	
                            else	
                            {	
                                pen = new Pen(SystemColors.ControlLightLight);	
                            }
```
...i.e., if `color == SystemColors.Control` then `pen` should be `new Pen(SystemColors.ControlLightLight)`

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs (hunk 10 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:8457]: Whoops, good catch :)

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 13]
```diff
@@ -670,6 +670,9 @@ protected override void OnRightToLeftChanged(EventArgs e)
 
         protected override void OnPaintBackground(PaintEventArgs e)
         {
+            if (e is null)
+                throw new ArgumentNullException(nameof(e));
+
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs:675]: It threw already, I'm just getting it thrown up sooner.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs (hunk 9 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs:28]: Sure

File: src/System.Windows.Forms/src/System/Windows/Forms/Internal/Gdi/FontCache.Data.cs (hunk 8 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Internal/Gdi/FontCache.Data.cs:105]: Figured it was rare and not worth the complexity, particularly as we're caching this anyway.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewComboBoxCell.cs (hunk 12 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewComboBoxCell.cs:2173]: We aren't as that giant chunk of code above was a no-op. We _always_ had a system color (`stockColor`). Threw me for a loop when I went through it. 😆 


``` C#
                            Color color = SystemColors.Control;
                            Color buttonShadow;
                            Color buttonShadowDark;
                            Color buttonFace = color;
                            Color highlight;
                            bool stockColor = color.ToKnownColor() == SystemColors.Control.ToKnownColor();   // <-- Um, yeah, that's identity
                            bool highContrast = SystemInformation.HighContrast;
                            if (color == SystemColors.Control)
```

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        }
+
+        /// <summary>
+        ///  Returns a cached <see cref="Pen"/>. Use in a using and assign to var.
+        /// </summary>
+        /// <remarks>
+        ///  Correct: using var pen = GdiPlusCache.GetCachedPen(Color.Blue);
+        ///  Incorrect (LEAKS): using Pen pen = GdiPlusCache.GetCachedPen(Color.Blue);
+        /// </remarks>
+        internal static PenCache.Scope GetCachedPen(this Color color) => GetPenScope(color);
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs:69]:
```suggestion
        internal static PenCache.Scope GetCachedPenScope(this Color color) => GetPenScope(color);
```

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+        /// <summary>
+        ///  Returns a cached <see cref="Pen"/>. Use in a using and assign to var.
+        /// </summary>
+        /// <remarks>
+        ///  Correct: using var pen = GdiPlusCache.GetCachedPen(Color.Blue);
+        ///  Incorrect (LEAKS): using Pen pen = GdiPlusCache.GetCachedPen(Color.Blue);
+        ///
+        ///  Debug builds track proper disposal.
+        /// </remarks>
+        internal static PenCache.Scope GetCachedPen(this Color color, int width)
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs:80]:
```suggestion
        internal static PenCache.Scope GetCachedPenScope(this Color color, int width)
```

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        /// <summary>
+        ///  Returns a cached <see cref="SolidBrush"/>. Use in a using and assign to var.
+        /// </summary>
+        /// <remarks>
+        ///  Correct: using var pen = GdiPlusCache.GetCachedSolidBrush(Color.Blue);
+        ///  Incorrect (LEAKS): using Pen pen = GdiPlusCache.GetCachedSolidBrush(Color.Blue);
+        ///
+        ///  Debug builds track proper disposal.
+        /// </remarks>
+        internal static SolidBrushCache.Scope GetCachedSolidBrush(this Color color) => GetSolidBrushScope(color);
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/GdiPlus/GdiPlusCache.cs:92]:
```suggestion
        internal static SolidBrushCache.Scope GetCachedSolidBrushScope(this Color color) => GetSolidBrushScope(color);
```
